autogluon/utils/tabular/ml/models/tab_transformer/TabTransformer_model.py

```diff
@@ -0,0 +1,320 @@
+                return out, features
+
+            def fit(self, trainloader, valloader=None, state=None):
+                """
+                Main training function for TabTransformer
+                "state" must be one of [None, 'pretrain', 'finetune']
+                None: corresponds to purely supervised learning
+                pretrain: discirminative task will be a pretext task
+                finetune: same as superised learning except that the model base has 
+                          exponentially decaying learning rate.
+                """
+
+                pretext_tasks=pretexts.__dict__
+                optimizers=[]
+                lr=self.kwargs['tab_kwargs']['lr']
+                weight_decay=self.kwargs['tab_kwargs']['weight_decay']
+                if state==None:
+                    optimizers = [optim.Adam(self.parameters(), lr=lr, weight_decay=weight_decay)]
+                    pretext=pretext_tasks['SUPERVISED_pretext'](self.kwargs)
+                elif state=='pretrain':
+                    optimizers = [optim.Adam(self.parameters(), lr=lr, weight_decay=weight_decay)]
+                    pretext=pretext_tasks['BERT_pretext'](self.kwargs)
+                elif state=='finetune':
+                    base_exp_decay=self.kwargs['tab_kwargs']['base_exp_decay']
+                    optimizer_fc    = optim.Adam(self.fc.parameters(), lr=lr, weight_decay=weight_decay) 
+                    optimizer_embeds = optim.Adam(self.embed.parameters(), lr=lr, weight_decay=weight_decay) 
+                    scheduler = optim.lr_scheduler.ExponentialLR(optimizer_embeds,gamma=base_exp_decay)
+                    optimizers.append(optimizer_fc)
+                    optimizers.append(optimizer_embeds)
+
+                    pretext=pretext_tasks['SUPERVISED_pretext'](self.kwargs)
+
+                else:
+                    raise NotImplementedError("state must be one of [None, 'pretrain', 'finetune']")
+
+                if self.kwargs['problem_type']=='regression':
+                    loss_criterion = nn.MSELoss()
+                else:
+                    loss_criterion = nn.CrossEntropyLoss()
+
+                old_val_accuracy=0.0
+         
+                epochs = self.kwargs['pretrain_epochs'] if state=='pretrain' else self.kwargs['epochs']
+                freq   = self.kwargs['pretrain_freq'] if state=='pretrain' else self.kwargs['freq']
+   
+                for e in range(1,epochs+1):
+                    _ = utils.epoch(self, trainloader, optimizers, loss_criterion=loss_criterion, \
+                                        pretext=pretext, state=state, scheduler=None, epoch=e, epochs=epochs, aug_kwargs=self.kwargs['augmentation']) #returns train_loss, train_acc@1, train_acc@5 
+                    
+                    if valloader is not None:
+                        if e % freq == 0:
+                            _, val_accuracy = utils.epoch(self, valloader, optimizers=None, \
```

Comment:
Note that the user may want to target a different metric than accuracy.
Instead of this, simply call `self.score` after each epoch as demonstrated here:
https://github.com/awslabs/autogluon/blob/master/autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py#L338
(higher scores will correspond to better performance)